Spurious "'catch' block is unreachable" warning for `_ = try` in a skipped local function body

### Description

Swift incorrectly diagnoses a `catch` block as unreachable when the corresponding `do` block explicitly discards a throwing call using `_ = try` inside a local function nested within top-level code.

The warning occurs when non-inlinable function bodies are skipped during type checking. The Swift driver enables this optimization while generating a module, so the warning is emitted by `swiftc -emit-module` and during a SwiftPM build, but not by a direct `swiftc -typecheck` invocation without the function-body-skipping option.

The throwing function's return type is not the determining factor. The warning occurs with both `Void` and non-`Void` return types when the call is written as `_ = try ...`.

This appears to be a Swift 6 compiler regression. I can reproduce it with Swift 6.0.3 through recent Swift 6.5 development snapshots, but not with Swift 5.10.1.

### Reproduction

The issue can be reproduced with a local function declared inside a top-level `do` statement. Save the following as `main.swift`:

```swift
func someFunc() throws -> Int { 0 }

do {
  func run() {
    do {
      _ = try someFunc()
    } catch {
      // ...
    }
  }
}
```

It also occurs when the local function is declared inside an immediately invoked global closure:

```swift
func someFunc() throws -> Int { 0 }

let closure: Void = {
  func run() {
    do {
      _ = try someFunc()
    } catch {
      // ...
    }
  }

  run()
}()
```

The syntactic form `_ = try ...`, rather than the throwing function's return type, is the relevant trigger. A non-`Void` throwing function produces the incorrect warning:

```swift
func someFunc() throws -> Int { 0 }

do {
  func run() {
    do {
      _ = try someFunc()
    } catch {  // <--- warning: 'catch' block is unreachable because no errors are thrown in 'do' block
        // ...
    }
  }
}
```

The same warning is emitted when the throwing function returns `Void`, as long as the call is explicitly discarded with `_ =`:

```swift
func someFunc() throws -> Void {}

do {
  func run() {
    do {
      _ = try someFunc()
    } catch {  // <--- warning: 'catch' block is unreachable because no errors are thrown in 'do' block
        // ...
    }
  }
}
```

Conversely, using a discard pattern binding instead of a discard assignment avoids the incorrect unreachable-`catch` warning, even when the function returns a non-`Void` value:

```swift
func someFunc() throws -> Int { 0 }

do {
  func run() {
    do {
      let _ = try someFunc()
    } catch {  // <--- No unreachable-'catch' warning
        // ...
    }
  }
}
```

Run:

```console
% swiftc -emit-module main.swift

main.swift:7:11: warning: 'catch' block is unreachable because no errors are thrown in 'do' block
 5 |         do {
 6 |             _ = try someFunc()
 7 |         } catch {
   |           `- warning: 'catch' block is unreachable because no errors are thrown in 'do' block
 8 |             // ...
 9 |         }
```

The warning is also emitted by a SwiftPM executable target containing the same source:

```console
% swift build

Building for debugging...
/Users/treastrain/Developer/IncorrectWarningInDoCatchPlayground/Sources/main.swift:7:11: warning: 'catch' block is unreachable because no errors are thrown in 'do' block
 5 |         do {
 6 |             _ = try someFunc()
 7 |         } catch {
   |           `- warning: 'catch' block is unreachable because no errors are thrown in 'do' block
 8 |             // ...
 9 |         }
[7/7] Applying IncorrectWarningInDoCatchPlayground
Build complete! (0.44s)
```

### Expected behavior

No unreachable-`catch` warning should be emitted because `someFunc()` is declared `throws` and is called with `try` inside the `do` block.

### Environment

Reproduced with:

- Swift version 6.5-dev (`LLVM 6f2057ffeafd4c6`, `Swift 83c32e02f71e4bb`)
- Swift version 6.4-dev (`LLVM 264fd65923c28d9`, `Swift ef761e567dc94ee`)
- Swift version 6.3.3 (`swift-6.3.3-RELEASE`)
- Swift version 6.2.4 (`swift-6.2.4-RELEASE`)
- Swift version 6.1.3 (`swift-6.1.3-RELEASE`)
- Swift version 6.0.3 (`swift-6.0.3-RELEASE`)

I reproduced the issue on at least the following targets:

- `arm64-apple-macosx26.0`
- `aarch64-unknown-linux-gnu`

Not reproduced with:

- Swift version 5.10.1 (`swift-5.10.1-RELEASE`)

### Additional information

- The warning is emitted in both Swift 5 and Swift 6 language modes.
- `swiftc -typecheck` without function-body skipping does not emit the warning for this source; `swiftc -emit-module` does.
- The module-emission frontend invocation adds `-experimental-skip-non-inlinable-function-bodies-without-types`. Adding this option to a type-check invocation reproduces the warning, while removing it from an otherwise equivalent `swift-frontend -emit-module` invocation avoids the warning.
- Both the top-level `do` form and the immediately invoked global closure form shown above reproduce the same incorrect warning.
- Immediate invocation is not required: a local function declared inside a non-invoked global closure initializer also reproduces the warning.
- The return type is not causal. `_ = try someFunc()` reproduces the warning for both `throws -> Int` and `throws -> Void`.
- The exact discard syntax matters: `_ = try someFunc()` reproduces the warning, while `let _ = try someFunc()` does not. A bare `try someFunc()` also does not produce the unreachable-`catch` warning.
- Declaring the function directly at global scope, moving the nested function into another function, or placing the throwing call directly in the top-level statement avoids the warning.
- Compiler investigation points to the interaction between function-body skipping and local-function effects checking. A skipped local function body remains parsed but is not pre-checked or type-checked. In that state, `_ = try someFunc()` is still represented by an unresolved `SequenceExpr`; effects checking stops at that unchecked expression and fails to record the throwing call.

--- a/lib/Sema/TypeCheckEffects.cpp
+++ b/lib/Sema/TypeCheckEffects.cpp
@@ -5031,6 +5031,9 @@ void TypeChecker::checkFunctionEffects(AbstractFunctionDecl *fn) {
   PrettyStackTraceDecl debugStack("checking effects handling for", fn);
 #endif
 
+  if (fn->isBodySkipped())
+    return;
+
   auto context = Context::forFunction(fn);
   auto &ctx = fn->getASTContext();
   CheckEffectsCoverage checker(ctx, context);

--- a/test/SILGen/skip_local_function_bodies.swift
+++ b/test/SILGen/skip_local_function_bodies.swift
@@ -1,6 +1,8 @@
 // RUN: %target-swift-emit-silgen-ossa -o /dev/null -enable-sil-opaque-values %s
-// RUN: %empty-directory(%t)
-// RUN: %target-swift-frontend -emit-module -experimental-skip-non-inlinable-function-bodies-without-types %s -emit-module-path %t/skip_local_function_bodies.swiftmodule
+// RUN: %target-swift-frontend -emit-module -verify %s -emit-module-path /dev/null
+// RUN: %target-swift-frontend -emit-module -verify -experimental-skip-non-inlinable-function-bodies-without-types %s -emit-module-path /dev/null
+// RUN: %target-swift-frontend -emit-module -verify -experimental-skip-non-inlinable-function-bodies %s -emit-module-path /dev/null
+// RUN: %target-swift-frontend -emit-module -verify -experimental-skip-all-function-bodies %s -emit-module-path /dev/null
 
 public protocol P {
     init()
@@ -11,7 +13,32 @@ extension P {
         typealias T = Self
 
         func g(_ x: T) {}
-	g(self)
+        g(self)
     }
 }
 
+func throwsError() throws -> Int { 0 }
+
+extension P {
+    public func h() {
+        typealias T = Self
+
+        func g(_ x: T) {
+            do {
+                _ = try throwsError()
+            } catch {
+            }
+        }
+        g(self)
+    }
+}
+
+let globalWithLocalFunc: Void = {
+    func g() {
+        do {
+            _ = try throwsError()
+        } catch {
+        }
+    }
+    g()
+}()

